vm/src/stdlib/random.rs

```diff
@@ -1,46 +1,138 @@
-use rand_distr::Normal;
+use std::cell::RefCell;
 
-use crate::pyobject::{PyObjectRef, PyResult};
-use crate::vm::VirtualMachine;
+use num_bigint::{BigInt, Sign};
+use num_traits::Signed;
+use rand::RngCore;
 
-pub fn make_module(vm: &VirtualMachine) -> PyObjectRef {
-    let ctx = &vm.ctx;
+use crate::function::OptionalOption;
+use crate::obj::objtype::PyClassRef;
+use crate::obj::objint::PyIntRef;
+use crate::pyobject::{PyClassImpl, PyObjectRef, PyRef, PyResult, PyValue};
+use crate::VirtualMachine;
 
-    py_module!(vm, "random", {
-        "gauss" => ctx.new_function(random_normalvariate), // TODO: is this the same?
-        "normalvariate" => ctx.new_function(random_normalvariate),
-        "random" => ctx.new_function(random_random),
-        // "weibull", ctx.new_function(random_weibullvariate),
-    })
+mod mersenne;
+
+#[derive(Debug)]
+enum PyRng {
+    Std(rand::rngs::ThreadRng),
+    MT(mersenne::MT19937),
+}
+
+impl Default for PyRng {
+    fn default() -> Self {
+        PyRng::Std(rand::thread_rng())
+    }
+}
+
+impl RngCore for PyRng {
+    fn next_u32(&mut self) -> u32 {
+        match self {
+            Self::Std(s) => s.next_u32(),
+            Self::MT(m) => m.next_u32(),
+        }
+    }
+    fn next_u64(&mut self) -> u64 {
+        match self {
+            Self::Std(s) => s.next_u64(),
+            Self::MT(m) => m.next_u64(),
+        }
+    }
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        match self {
+            Self::Std(s) => s.fill_bytes(dest),
+            Self::MT(m) => m.fill_bytes(dest),
+        }
+    }
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand::Error> {
+        match self {
+            Self::Std(s) => s.try_fill_bytes(dest),
+            Self::MT(m) => m.try_fill_bytes(dest),
+        }
+    }
+}
+
+#[pyclass(name = "Random")]
+#[derive(Debug)]
+struct PyRandom {
+    rng: RefCell<PyRng>,
+}
+
+impl PyValue for PyRandom {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.class("_random", "Random")
+    }
 }
 
-fn random_normalvariate(mu: f64, sigma: f64, vm: &VirtualMachine) -> PyResult<f64> {
-    let normal = Normal::new(mu, sigma).map_err(|rand_err| {
-        vm.new_exception_msg(
-            vm.ctx.exceptions.arithmetic_error.clone(),
-            format!("invalid normal distribution: {:?}", rand_err),
-        )
-    })?;
-    let value = normal.sample(&mut rand::thread_rng());
-    Ok(value)
+#[pyimpl(flags(BASETYPE))]
+impl PyRandom {
+    #[pyslot(new)]
+    fn new(cls: PyClassRef, vm: &VirtualMachine) -> PyResult<PyRef<Self>> {
+        PyRandom {
+            rng: RefCell::new(PyRng::default()),
+        }
+        .into_ref_with_type(vm, cls)
+    }
+
+    #[pymethod]
+    fn random(&self) -> f64 {
+        gen_res53(&mut *self.rng.borrow_mut())
+    }
+
+    #[pymethod]
+    fn seed(&self, n: OptionalOption<PyIntRef>) {
+        let new_rng = match n.flat_option() {
+            None => PyRng::default(),
+            Some(n) => {
```

Comment:
I would like to suggest that instead of using `MT` when being called with seed we should keep using the `Std` with [SeedableRng](https://rust-num.github.io/num/rand/trait.SeedableRng.html). I think that users should receive the best available random as promised by `StdRng`.
I would suggest the following in order to allow using `MT`. Add the `version` parameter with a default of `3` and when this is called with `1` or `2` fallback to `MT`.

Reply:
The version parameter is entirely managed in `random.py`; it only changes how strings and bytes are turned into seeds. The entire point of the mersenne twister implementation is to stay compatible with CPython's seeding, and if we don't do that then it's kind of pointless. `random` isn't meant to be cryptographically secure anyway, from the `random` docs: "The pseudo-random generators of this module should not be used for security purposes. For security or cryptographic uses, see the secrets module". I'd say that supplying actual random numbers when the seed is `None` is more than enough.